File: py/1090.largest-values-from-labels.py
#
# @lc app=leetcode id=1090 lang=python3
#
# [1090] Largest Values From Labels
#
from collections import Counter
# @lc code=start


class Solution:
    def largestValsFromLabels(self, values: List[int], labels: List[int], numWanted: int, useLimit: int) -> int:
        decreasing_idx = list(range(len(values)))
        decreasing_idx.sort(key=lambda x: -values[x])
        used = Counter()
        res = 0
        for idx in decreasing_idx:
            if numWanted == 0:
                break
            used[labels[idx]] += 1
            if used[labels[idx]] <= useLimit:
                res += values[idx]
                numWanted -= 1
        return res
# @lc code=end

File: 1090.largest-values-from-labels.cpp
/*
 * @lc app=leetcode id=1090 lang=cpp
 *
 * [1090] Largest Values From Labels
 */
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <numeric>
using namespace std;
// @lc code=start
class Solution
{
public:
    int largestValsFromLabels(vector<int> &values, vector<int> &labels, int numWanted, int useLimit)
    {
        int w = values.size();
        vector<int> decreasing_ldx(w);
        iota(decreasing_ldx.begin(), decreasing_ldx.end(), 0);
        sort(decreasing_ldx.begin(), decreasing_ldx.end(), [&](int l, int r)
             { return values[l] > values[r]; });
        unordered_map<int, int> used;
        int res = 0;
        for (auto idx : decreasing_ldx)
        {
            if (numWanted == 0)
            {
                break;
            }
            if (++used[labels[idx]] <= useLimit)
            {
                res += values[idx];
                --numWanted;
            }
        }
        return res;
    }
};
// @lc code=end
